plugins/targetlocker/mysql/mysql_target_locker.go

```diff
@@ -0,0 +1,448 @@
+) (result sql.Result, err error) {
+	// Remove expired an lock (if exists)
+
+	_, _ = q.actionClearExpiredLock(mysql, targetID)
+
+	// Add the lock
+
+	var errs []error
+
+	result, execErr := q.exec(mysql,
+		"INSERT INTO `"+q.tableName()+"` "+
+			"SET target_id=?, job_id=?, created_at = NOW(), expires_at = NOW() + INTERVAL ? SECOND",
+		targetID, jobID, int64(q.lockTTL.Seconds())) // #nosec G202
+
+	// Check errors:
+
+	if execErr != nil {
+		errs = append(errs, execErr)
+	}
+
+	if result != nil {
+		_, rowsAffectedErr := result.RowsAffected()
+		if rowsAffectedErr != nil {
+			errs = append(errs, rowsAffectedErr)
+		}
+	}
+
+	for _, errItem := range errs {
+		if q.errorType(errItem) == mysqlErrorTypeDuplicateKey {
+			err = targetlocker.ErrAlreadyLocked{
+				TargetID:        targetID,
+				UnderlyingError: errItem,
+			}
+		}
+	}
+
+	return
+}
+
+func (q mysqlLocksQuerier) actionUnlock(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetID string,
+) (sql.Result, error) {
+	return q.exec(mysql,
+		"DELETE FROM `"+q.tableName()+"` "+
+			"WHERE target_id=? AND job_id=?", targetID, jobID) // #nosec G202
+}
+
+func (q mysqlLocksQuerier) actionRefreshLock(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetID string,
+) (sql.Result, error) {
+	return q.exec(mysql,
+		"UPDATE `"+q.tableName()+"` "+
+			`SET expires_at = NOW() + INTERVAL ? SECOND WHERE target_id=? AND job_id=?`,
+		int64(q.lockTTL.Seconds()), targetID, jobID) // #nosec G202
+}
+
+func (q mysqlLocksQuerier) placeholder(arg interface{}) string {
+	switch arg := arg.(type) {
+	case []string:
+		if len(arg) == 0 {
+			return `()`
+		}
+		result := strings.Repeat(`?,`, len(arg))
+		return "(" + result[:len(result)-1] + ")"
+	default:
+		panic(fmt.Sprintf("unknown argument type: %T", arg))
+	}
+}
+
+func (q mysqlLocksQuerier) arguments(arg interface{}) (result []interface{}) {
+	switch arg := arg.(type) {
+	case []string:
+		for _, item := range arg {
+			result = append(result, item)
+		}
+	default:
+		panic(fmt.Sprintf("unknown argument type: %T", arg))
+	}
+	return
+}
+
+func (q mysqlLocksQuerier) getLocks(
+	mysql backendQuerier,
+	jobID types.JobID,
+	targetIDs []string,
+) (*sql.Rows, error) {
+	arguments := []interface{}{jobID}
+	arguments = append(arguments, q.arguments(targetIDs)...)
+	return q.query(mysql,
+		"SELECT `target_id` FROM `"+q.tableName()+"` "+
+			"WHERE job_id=? AND target_id IN "+q.placeholder(targetIDs),
+		arguments...) // #nosec G202
+}
+
+func (tl *TargetLocker) reportIfError(err error) {
+	if err == nil {
+		return
+	}
+	frame := runtimetools.Frame(1)
+	tl.log.Errorf("%v:%v:%v: %v", frame.File, frame.Line, frame.Function, err)
+}
+
+func (tl *TargetLocker) tx(description string, fn func(tx *sql.Tx) error) (err error) {
+	tx, err := tl.mySQL.Begin()
+	if err != nil {
+		return fmt.Errorf("unable to '%s': unable to start transaction: %w", description, err)
+	}
+
+	tl.log.Debugf("Start transaction -> %p", tx)
+
+	defer func() {
+		if err != nil {
```

Comment:
It will have the value assigned by `fn(tx)` below.

Reply:
Well, ok, that is hard to follow. But it seems to be an idiomatic way to take actions on the error value returned by a function, so I'll accept that it's only me who is not familiar with some idioms enough.

Reply:
I still think that trading off for readability for a reader that is not expert in all the possible Golang idioms is the preferred choice, and here I think not deferring error checking would be a tradeoff in better readability. What do you think?

Reply:
Personally I don't like to put complex logic directly to `defer` (usually it's better to declare a separate function for that and put it to the `defer`). But I think that receiving and changing the returned value through `defer` is a standard Go's feature, so it should be fine to use it sometimes.

On the other hand, it's a "plugin" and may be you right: a ConTest user (not developer) may become a potential reader of this code. OK, I'll try to avoid using this feature in here. Will be fixed.

Reply:
Refactored the function.